### ISSUE TITLE
Variant incorrectly displaying unit name not display as in shopfront and backoffice

### ISSUE THREAD

**Issue by chezaorchard** [openfoodfoundation/openfoodnetwork#8553]
**Description**
Product variants, where unit size=item, are displaying in all open shopfronts with only the unit name, not the full 'display as' which usually includes a number as well as descriptor. 

Product variants, where unit size=item, are displaying in customer totals report with only the unit name, not the full 'display as' 

**Investigation:**
- Affecting the backoffice (reports), and the customer shopfront experience
- Only where unit size=item
- Happening for accounts where sells = all OR sells = any
- Affecting customer totals reports, but pack by reports have the correct 'display as' in variant column
- Displaying incorrectly in backoffice orders, but displaying correctly on order confirmation email 
- Issue appeared at some point after approx 10pm AEST December 1 (determined by download time of user where the report did include 'display as' but when run today it is only displaying 'unit name')

Tested with user IDs:
2840
1927
153
2246
639

**Expected Behavior**
All product variants should display in shopfronts and reports with complete 'display as' value

**Actual Behaviour**
Product variants are incorrectly displaying in shopfronts and some reports as 'unit name' only

**Steps to Reproduce**
Navigate to open shop
Scroll down to product variant where size=item
Observe variant is incorrectly displaying with 'unit name' only

Log in as superadmin (User ID 2840)
Run customer totals report for current order cycle
Observe column G incorrectly displays 'unit name' only

Long in as customer (User ID 2840)
Run customer totals report for current order cycle
Observe column G incorrectly displays 'unit name' only

Check with User ID 2840 via phone - they downloaded the report yesterday evening (approx Australian time 10pm, Wed Dec) and the report did contact the correct 'display as' at that time. 

Screenshot

<img width="1041" alt="Screen Shot 2021-12-02 at 4 06 58 pm" src="https://user-images.githubusercontent.com/68367600/144364640-eed8b6b2-7c42-4284-b5fa-53bc07e0fc04.png">

<img width="1158" alt="Screen Shot 2021-12-02 at 4 09 01 pm" src="https://user-images.githubusercontent.com/68367600/144364642-9966969d-b813-4699-8876-3f99f4a272c9.png">

<img width="911" alt="Screen Shot 2021-12-02 at 4 25 19 pm" src="https://user-images.githubusercontent.com/68367600/144364653-3fd7a8ff-582a-42e2-86e7-e9bf6875b7e5.png">

<img width="1376" alt="Screen Shot 2021-12-02 at 4 25 33 pm" src="https://user-images.githubusercontent.com/68367600/144364662-12a3c4c2-0509-4b53-9ab3-a8699c2dcff3.png">

<img width="1392" alt="Screen Shot 2021-12-02 at 4 33 51 pm" src="https://user-images.githubusercontent.com/68367600/144364875-52e90642-2809-4ec3-9068-4abcc4ef8c2c.png">

<img width="636" alt="Screen Shot 2021-12-02 at 4 33 19 pm" src="https://user-images.githubusercontent.com/68367600/144364883-3da0a959-9d36-4383-90bb-16adf0dfe15c.png">

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Browser name and version: Chrome Version 96.0.4664.55 (Official Build) (x86_64)
* Operating System and version (desktop or mobile): Desktop & Mobile

**Comment by lin-d-hop** [comment]
Is this related to #8500?

**Comment by RachL** [comment]
@lin-d-hop Judging by Chez comment on slack it seems #8500 introduced this problem :(

**Comment by lin-d-hop** [comment]
@apricot12 Can we assign to you in this case?
@openfoodfoundation/testers Let's also make sure we take both cases into account during testing.

**Comment by apricot12** [comment]
@lin-d-hop Yes, I'll look at it immediately.

**Comment by jibees** [comment]
> @openfoodfoundation/testers Let's also make sure we take both cases into account during testing.

Could we also create an automatic test for this?

**Comment by RonellaG** [comment]
@apricot12 @jibees @lin-d-hop @RachL We've started to have more hubs enquire about this issue. This is impacting their shopfront listings as well as their ability to generate packing reports. What is the status of this bug and how soon can it be resolved? Given it impacts our hubs sales and operations quite a lot can it please be labelled S1?

**Comment by chezaorchard** [comment]
User ID 286, a bulk food hub, are experiencing this issue with their current order cycle: "The pricing for e.g. coconut cream is $3.25/can if bought by the can or in 6's but $2.55/can if ordered in 12's. However the shopper will not realise how many cans they are buying as the number has now disappeared."

**Comment by apricot12** [comment]
> @apricot12 @jibees @lin-d-hop @RachL We've started to have more hubs enquire about this issue. This is impacting their shopfront listings as well as their ability to generate packing reports. What is the status of this bug and how soon can it be resolved? Given it impacts our hubs sales and operations quite a lot can it please be labelled S1?

Hey, It should be up for code review by today. I'll try to speed it up as much as possible :)

**Comment by apricot12** [comment]
@chezaorchard @RonellaG I've just put up a [PR](https://github.com/openfoodfoundation/openfoodnetwork/pull/8572) that fixes this issue. Now it's just a matter of going through the review and testing process and getting it into this weeks release.

**Comment by Matt-Yorkley** [comment]
Hey :wave:, I've just been looking through this and the other related issues (#8495) and recent changes (#8500).

It's a tricky one! The way these different values are saved is really complex in terms of the datamodel, and the code that handles it is quite arcane (the `VariantUnits::VariantAndLineItemNaming` service). The tricky part is that even small changes in that service can potentially have lots of different effects all throughout the codebase, and it's almost impossible to predict what they'll be.

I think I would suggest maybe reverting the changes in #8500, and then trying to re-assess how to solve the original issue #8495 by making some changes in the view files that handle displaying products in the shopfront, eg:

- https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/views/shop/products/_summary.html.haml
- https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/assets/javascripts/templates/shop_variant.html.haml

**Comment by apricot12** [comment]
@Matt-Yorkley I was wondering the same thing. I wanted to try implementing some view logic to fix this, but I wasnt sure if it was the right thing to do. It seemed urgent too so I just went with a hacky fix.

**Comment by Matt-Yorkley** [comment]
> I wasnt sure if it was the right thing to do.

Fair enough! I'm personally terrified of touching that service :sweat_smile: It's really difficult to work with.

**Comment by mkllnk** [comment]
It sounds like a good idea to replace its use bit by bit then. For another PR I went through the code just to find out which database fields it was using and it which order and that was an essay already.

I just reverted #8500 on au-prod because it's causing huge problems for our enterprises here. We should probably do that in master and then any further change needs to first add a spec for this issue (shopfront and reports were affected) and then find a better way of solving the original issue.
